design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+# Proposal: Define requirements new traits to qualify and enhance the Conan dependencies model
+
+
+| **Status**        |                                                   |
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [026](https://github.com/conan-io/tribe/pull/26)  |
+| **Submitted**     | 2021-07-06                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+Introduce the concept of "requirements traits", to better model the relations between packages. Requirement traits are specifiers of the Conan ``requires`` that can be defined in recipes. These traits can allow defining the current Conan 1.X ``build_requires`` behavior with some ``requires`` traits.
+
+This proposal does not affect ``python_requires``, because of their nature and evaluation at recipe parsing time, they are very different to regular ``requires`` and ``build_requires``, and they would not benefit from this proposal.
```

Comment:
It is much different to have a ``build_requires`` containing some python code, that can be used to build, than a ``python_requires``. There are packages out there, like https://conan.io/center/meson which is a python code, that can be used as build_requires.

``python_requires`` is much much different, because it needs to be evaluated at recipe parsing time, which happens way before any package binary is fetched. So all the traits that relate to binaries: headers, libs, run, transitive_headers, transitive_libs, directly not apply.

So ``python_requires`` are not in the build context, and they are not in the host context. The build and host contexts are settings, and ``python_requires`` do not have settings, because they do not have binaries. They are python code that becomes part of the recipe that calls them, without any context.

The reason here could be that the tool you are requiring is more than just some python code to be reused, but a complete build tool that is wrapped in python, and this is the reason why it cannot be put inside a ``python_requires``?
I think the best line of action for this case might be to try to use a thin layer over such a tool, and launch it as a subprocess, instead of ``import``ing it directly. Specially if you can have different versions of that tool in your dependency graph. 

The python packaging and importing ecosystem is also complicated, and Conan was not designed to solve that problem, so this would be very difficult to solve: basically you are trying to ``import`` the same modules, different versions from different locations in the same python interpreter, something Python was not designed for, so this would be challenging for any tool out there, not only Conan.

Reply:
> The build and host contexts are settings, and python_requires do not have settings, because they do not have binaries. 

Ah. I admit I have not actually written much using `python_requires`, and I had managed to overlook that they did not have settings at all . Which would seem to imply they also simply don't carry transitive requires either? Given all that the statement that "they would not benefit from this proposal" makes more sense.

I was thinking of examples like https://docs.conan.io/en/latest/extending/custom_build_helper.html (which recommends the use of python_requires) to obtain WafBuildEnvironment, and thinking (as a what-if) of such a build helper that was either contained or required a binary package that actually contained the waf build tool itself.

But it sounds like that is a categorical error - if one wanted to do that, you would need both a python_requires and a build_requires (which might, or might not, be the same reference, but are loaded at very different ways). And it seems to be legal to have both python_requires and build_requires for the same reference. That's only [documented in the context of an unrelated workaround](https://docs.conan.io/en/latest/devtools/build_requires.html#making-build-requires-affect-the-consumers-package-id), but here it would be simply be the right thing to do - the python_requires to get the build helper (in no context) and a build_requires to get the tool (in build context).

Thank you for the explanation, that hopefully makes my mental model of what python_requires are for and how they work a bit more accurate.

> basically you are trying to import the same modules, different versions from different locations in the same python interpreter, something Python was not designed for

I concur, but it seemed like this scoping/versioning was was at least part of what `python_requires` was trying to solve, and https://docs.conan.io/en/latest/howtos/python_code_reuse.html says 

> To reuse Python code, from Conan 1.7 there is a new python_requires() feature. See: Python requires: reusing Python code in recipes This “how to” might be deprecated and removed in the future. It is left here for reference only.

So it seemed like it might be intended that python_requires evolve to replace such usage. But in any I now see why that still wouldn't create a reason to label python_requires as `build`/`test`/etc, and thus wouldn't be directly related to this proposal.

Reply:
> The only way to achieve this level of isolation and independence, via a very aggressive configuration/settings erasure, would be the concept of "re-packaging", 

Yes,that's basically what we do now - there are commonly two recipes, one for obtaining dependencies and building the binaries, and an entirely different one (to be used used with export-pkg) to copy things into the cache for upload consumption or upload. The latter simply fails to mention the majority of the settings/requirements (omitting all those which would be `visible=False` and are not important to a consumer). This course has lots of drawbacks: there's duplication that one can mess up accidentally not synchronize when changing the two recipes, one can never use `conan create` or `--build missing`, and the conaninfo.txt entirely fails to record what requirements/settings the uploaded binary really is (even in the `[full_requires]/`[full_settings]` that would normally know about things that were excluded from package_id). 

Hence I'd love to see things `visible=False` and `compatible_packages` become adequate for this kind of library :-)

Reply:
I'm on the same page as @memsharded: what I've successfully used, 3 parts:
1. A Toolchain or Deps class (former generator) - takes conanfile in constructor.
2. A Mixin to instantiate Toolchain class and call it in `generate()` method. Also provides `build_requirements()` method to inject a dependency on the build tool.
3. The tool itself.

Former 2 are part of `python_require` and is a pure Python code (strictly a mixin can be skipped). The latter is a ordinary requirement intended to be used in the build context. May have settings.

The trick is all mixins need to play nicely together and call `super()`.

Reply:
Yes, what @ytimenkov is saying sounds a reasonable approach, that might scale and be reasonable to maintain. The truth is that so far we haven't found any blocker for using the same reference as ``python_require`` and regular ``requires`` or ``build_requires``, and while seems a bit weird to use the same exact reference for both things, it also seems to work, thanks to the fact that the resolution, context, etc are completely decoupled.

We will probably try to come up with more insights about if it is a good practice or not to use the same reference.